core/src/main/java/com/netflix/iceberg/ReplaceFiles.java

```diff
@@ -16,21 +16,113 @@
 
 package com.netflix.iceberg;
 
-import java.util.List;
-import java.util.Set;
+import com.netflix.iceberg.exceptions.CommitFailedException;
+import com.netflix.iceberg.exceptions.RuntimeIOException;
+import com.netflix.iceberg.exceptions.ValidationException;
+import com.netflix.iceberg.io.OutputFile;
+
+import java.io.IOException;
+import java.util.*;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
 
 class ReplaceFiles extends SnapshotUpdate implements RewriteFiles {
-  public ReplaceFiles(TableOperations ops) {
-    super(ops);
-  }
 
-  @Override
-  protected List<String> apply(TableMetadata base) {
-    return null;
-  }
+    private final TableOperations ops;
+    private final Set<DataFile> filesToAdd = new HashSet<>();
+    private final Set<CharSequence> pathsToDelete = new HashSet<>();
+    private final List<String> newManifests = new ArrayList<>();
+    private List<String> cachedManifests = new ArrayList<>();
+    private final AtomicInteger manifestCount = new AtomicInteger(0);
+    private boolean hasChanges = false;
+
+    public ReplaceFiles(TableOperations ops) {
+        super(ops);
+        this.ops = ops;
+    }
+
+    /**
+     * @param base TableMetadata of the base snapshot
+     * @return list of manifests that "may" get committed if commit is called on this instance.
+     */
+    @Override
+    protected List<String> apply(TableMetadata base) {
+        ValidationException.check(!(this.pathsToDelete.isEmpty() || this.filesToAdd.isEmpty()), "Must provide some files to add and delete.");
```

Comment:
`ValidationException` isn't intended for argument validation. Instead, I would suggest using `Preconditions` to throw `IllegalArgumentException`.

`ValidationException` is used for consistency checks, like making sure that a partition spec references valid column ids in a table schema.

Reply:
Changed.